File: stacks/ltcd_234.py
'''
LeetCode - problem 234
'''

    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow = fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        middle = slow
        prev, actual = None, middle
        while actual:
            nxt = actual.next
            actual.next = prev
            prev = actual
            actual = nxt
        l, r = head, prev
        while r:
            if l.val != r.val:
                return False
            l = l.next
            r = r.next
        return True

File: stacks/ltcd_234.cpp
//LeetCode - problem 234

    bool isPalindrome(ListNode* head) {
        ListNode *cur = head;//this is the normal cursor
        return judge(head, cur);
    }
    bool judge(ListNode *head, ListNode* &cur) {
        if (!head) return true;
        if (!judge(head->next, cur)) return false;
        if (cur->val != head->val) return false;
        else {cur = cur->next; return true;}
    }
